File: tests/julia/test_consumption_leisure.jl
using Test
using MacroAIComp

@testset "ConsumptionLeisure Tests" begin
    @testset "Equivalence fsolve and optim" begin
        params = default_calibration(ConsumptionLeisureParameters)
        W = fill(30.0, params.T)

        res_fsolve = solve_foc_fsolve(params, W)
        res_optim = solve_direct_optim(params, W)

        @test res_fsolve["C"] ≈ res_optim["C"] atol=1e-3
        @test res_fsolve["L"] ≈ res_optim["L"] atol=1e-3
        @test res_fsolve["B"] ≈ res_optim["B"] atol=1e-3
    end

    @testset "Terminal Condition" begin
        params = default_calibration(ConsumptionLeisureParameters)
        W = fill(30.0, params.T)

        res_fsolve = solve_foc_fsolve(params, W)
        res_optim = solve_direct_optim(params, W)

        @test abs(res_fsolve["B"][end]) < 1e-6
        @test abs(res_optim["B"][end]) < 1e-6
    end

    @testset "Leisure Consumption Tradeoff" begin
        params = default_calibration(ConsumptionLeisureParameters)
        W = fill(30.0, params.T)

        res = solve_foc_fsolve(params, W)

        @test all(l -> l >= 0.0, res["L"])
        @test all(l -> l < 1.0, res["L"])
        @test all(o -> o > 0.0, res["O"])
        @test all(o -> o <= 1.0, res["O"])
    end

    @testset "Preference Sensitivity" begin
        params_low = ConsumptionLeisureParameters(30, 0.97, 0.02, 0.40, 0.0)
        params_high = ConsumptionLeisureParameters(30, 0.97, 0.02, 0.60, 0.0)
        W = fill(30.0, params_low.T)

        res_low = solve_foc_fsolve(params_low, W)
        res_high = solve_foc_fsolve(params_high, W)

        mean_l_high = sum(res_high["L"]) / length(res_high["L"])
        mean_l_low = sum(res_low["L"]) / length(res_low["L"])

        @test mean_l_high > mean_l_low
    end

    @testset "Interest Rate Slope" begin
        params = ConsumptionLeisureParameters(30, 0.97, 0.05, 0.5, 0.0)
        W = fill(30.0, params.T)

        res = solve_foc_fsolve(params, W)

        for t in 1:(params.T - 1)
            @test res["C"][t + 1] > res["C"][t]
        end
    end
end

File: tests/python/test_consumption_leisure.py
import numpy as np
from macroaicomp.models.consumption_leisure import (
    ConsumptionLeisureParameters,
    solve_foc_fsolve,
    solve_direct_cvxpy,
)


def test_equivalence_fsolve_cvxpy():
    """Test that both fsolve and cvxpy solvers yield the same results."""
    params = ConsumptionLeisureParameters()
    W = np.full(params.T, 30.0)

    res_fsolve = solve_foc_fsolve(params, W)
    res_cvxpy = solve_direct_cvxpy(params, W)

    # Check that consumption, labor, and savings paths are identical within tolerances
    np.testing.assert_allclose(res_fsolve["C"], res_cvxpy["C"], rtol=1e-4, atol=1e-4)
    np.testing.assert_allclose(res_fsolve["L"], res_cvxpy["L"], rtol=1e-4, atol=1e-4)
    np.testing.assert_allclose(res_fsolve["B"], res_cvxpy["B"], rtol=1e-4, atol=1e-4)


def test_terminal_condition():
    """Test that the terminal stock of assets is exactly 0.0."""
    params = ConsumptionLeisureParameters()
    W = np.full(params.T, 30.0)

    res_fsolve = solve_foc_fsolve(params, W)
    res_cvxpy = solve_direct_cvxpy(params, W)

    assert abs(res_fsolve["B"][-1]) < 1e-6
    assert abs(res_cvxpy["B"][-1]) < 1e-6


def test_leisure_consumption_tradeoff():
    """Test that labor supply lies within rational boundaries (0 to 1)."""
    params = ConsumptionLeisureParameters()
    W = np.full(params.T, 30.0)

    res = solve_foc_fsolve(params, W)

    assert np.all(res["L"] >= 0.0)
    assert np.all(res["L"] < 1.0)
    assert np.all(res["O"] > 0.0)
    assert np.all(res["O"] <= 1.0)


def test_preference_sensitivity():
    """Test that a higher consumption weight (gamma) increases labor supply."""
    params_low = ConsumptionLeisureParameters(gamma=0.40)
    params_high = ConsumptionLeisureParameters(gamma=0.60)
    W = np.full(params_low.T, 30.0)

    res_low = solve_foc_fsolve(params_low, W)
    res_high = solve_foc_fsolve(params_high, W)

    # Average labor supply should be higher when consumption has a higher weight in utility
    assert np.mean(res_high["L"]) > np.mean(res_low["L"])


def test_interest_rate_slope():
    """Test that a high interest rate yields a positive consumption path slope."""
    # beta = 0.97, R = 0.05 => beta*(1+R) = 0.97 * 1.05 = 1.0185 > 1.0
    params = ConsumptionLeisureParameters(beta=0.97, R=0.05)
    W = np.full(params.T, 30.0)

    res = solve_foc_fsolve(params, W)

    # Consumption should have a positive slope
    for t in range(params.T - 1):
        assert res["C"][t + 1] > res["C"][t]
